cmd/ido2icingadb/main.go

```diff
@@ -0,0 +1,437 @@
+func fillCache() {
+	progress := mpb.New()
+	for _, ht := range types {
+		if ht.cacheFiller != nil {
+			ht.setupBar(progress)
+		}
+	}
+
+	types.forEach(func(ht *historyType) {
+		if ht.cacheFiller != nil {
+			ht.cacheFiller(ht)
+		}
+	})
+
+	progress.Wait()
+}
+
+// migrate does the actual migration.
+func migrate(c *Config, idb *icingadb.DB, envId []byte) {
+	endpointId := sha1.Sum([]byte(c.Icinga2.Endpoint))
+	idbTx := &sync.Mutex{}
+
+	progress := mpb.New()
+	for _, ht := range types {
+		ht.setupBar(progress)
+	}
+
+	types.forEach(func(ht *historyType) {
+		ht.migrate(c, idb, envId, endpointId, idbTx, ht)
+	})
+
+	progress.Wait()
+}
+
+// migrate does the actual migration for one history type.
+func migrateOneType[IdoRow any](
+	c *Config, idb *icingadb.DB, envId []byte, endpointId [sha1.Size]byte, idbTx *sync.Mutex, ht *historyType,
+	convertRows func(env string, envId, endpointId icingadbTypes.Binary,
+		selectCache func(dest interface{}, query string, args ...interface{}), ido *sqlx.Tx,
+		idoRows []IdoRow) (icingaDbInserts, icingaDbUpserts [][]any, checkpoint any),
+) {
+	var lastQuery string
+	var lastStmt *sqlx.Stmt
+
+	defer func() {
+		if lastStmt != nil {
+			_ = lastStmt.Close()
+		}
+	}()
+
+	selectCache := func(dest interface{}, query string, args ...interface{}) {
+		// Prepare new one, if old one doesn't fit anymore.
+		if query != lastQuery {
+			if lastStmt != nil {
+				_ = lastStmt.Close()
+			}
+
+			var err error
+
+			lastStmt, err = ht.cache.Preparex(query)
+			if err != nil {
+				log.With("backend", "cache", "query", query).
+					Fatalf("%+v", errors.Wrap(err, "can't prepare query"))
+			}
+
+			lastQuery = query
+		}
+
+		if err := lastStmt.Select(dest, args...); err != nil {
+			log.With("backend", "cache", "query", query, "args", args).
+				Fatalf("%+v", errors.Wrap(err, "can't perform query"))
+		}
+	}
+
+	var args map[string]interface{}
+
+	// For the case that the cache was older that the IDO,
+	// but ht.cacheFiller couldn't update it, limit (WHERE) our source data set.
+	if ht.cacheLimitQuery != "" {
+		var limit uint64
+		cacheGet(ht.cache, &limit, ht.cacheLimitQuery)
+		args = map[string]interface{}{"cache_limit": limit}
+	}
+
+	upsertProgress, _ := idb.BuildUpsertStmt(&IdoMigrationProgress{})
+
+	ht.bar.SetCurrent(ht.done)
+
+	// Stream IDO rows, ...
+	sliceIdoHistory(
+		ht.snapshot, ht.migrationQuery, args, ht.lastId,
+		func(idoRows []IdoRow) (checkpoint interface{}) {
+			// ... convert them, ...
+			inserts, upserts, lastIdoId := convertRows(
+				c.Icinga2.Env, envId, endpointId[:], selectCache, ht.snapshot, idoRows,
+			)
+
+			// ... and insert them:
+
+			if idb.DriverName() == driver.MySQL {
+				// Avoid MySQL error 1205 (Lock wait timeout exceeded; try restarting transaction)
+				// due to concurrent transactions upsert the same table (history).
+				idbTx.Lock()
+				defer idbTx.Unlock()
+			}
+
+			tx, err := idb.Beginx()
```

Comment:
Even though individual inserts are now batched into smaller chunks (a few hundred rows each), the size of this transaction is still determined by the number of rows read from the IDO at once (10k) where each row may generate multiple rows in Icinga DB (I think up to 3). So this transaction still takes very long (I've observed multiple minutes in my test setup), leading to long rollback times, which means https://github.com/Icinga/icingadb/pull/253#issuecomment-1161397199 https://github.com/Icinga/icingadb/pull/253#issuecomment-1161837886 is still present

Reply:
I've played around bit with this and on my machine with the IDO database from my test cluster (i.e. not the dump I posted in this issue), setting `bulk = 200` leads to transaction of about 4 seconds, so 100 seems to be the right order of magnitude.